In [ ]:



df=pd.read_csv('/datasets/credit_scoring_eng.csv')# Muat datanya


In [ ]:
df

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

In [ ]:
df.shape



(21525, 12)

In [ ]:
df.head()



children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [ ]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [ ]:

df[df['days_employed'].isnull()]


children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]


If we only look at the number of missing rows, which is 2174 when summed with 19351, the result is 21525, meaning it is symmetrical. But there are some data in the 'days_employed' column with strange values (minus and large), so it needs to be investigated further. 

In [ ]:

df.loc[(df['days_employed'].isna()) & (df['total_income'].isna())].shape


(2174, 12)

In [ ]:


df_null = df[(df['days_employed'].isnull()) & (df['total_income'].isnull())]
df_null.head(20)

children  days_employed  dob_years            education  education_id  \
12          0            NaN         65  secondary education             1   
26          0            NaN         41  secondary education             1   
29          0            NaN         63  secondary education             1   
41          0            NaN         50  secondary education             1   
55          0            NaN         54  secondary education             1   
65          0            NaN         21  secondary education             1   
67          0            NaN         52    bachelor's degree             0   
72          1            NaN         32    bachelor's degree             0   
82          2            NaN         50    bachelor's degree             0   
83          0            NaN         52  secondary education             1   
90          2            NaN         35    bachelor's degree             0   
94          1            NaN         34    bachelor's degree             0   
96          0            NaN         44  SECONDARY EDUCATION             1   
97          0            NaN         47    bachelor's degree             0   
120         0            NaN         46  secondary education             1   
121         0            NaN         29    bachelor's degree             0   
135         0            NaN         27  secondary education             1   
141         0            NaN         39  secondary education             1   
145         0            NaN         62  secondary education             1   
174         0            NaN         55    bachelor's degree             0   

         family_status  family_status_id gender    income_type  debt  \
12   civil partnership                 1      M        retiree     0   
26             married                 0      M  civil servant     0   
29           unmarried                 4      F        retiree     0   
41             married                 0      F  civil servant     0   
55   civil partnership                 1      F        retiree     1   
65           unmarried                 4      M       business     0   
67             married                 0      F        retiree     0   
72             married                 0      M  civil servant     0   
82             married                 0      F       employee     0   
83             married                 0      M       employee     0   
90             married                 0      F       employee     0   
94   civil partnership                 1      F       business     0   
96             married                 0      F       employee     0   
97             married                 0      F       employee     0   
120            married                 0      F       employee     0   
121            married                 0      F       business     0   
135            married                 0      M       business     0   
141  civil partnership                 1      M       employee     0   
145            married                 0      M        retiree     0   
174    widow / widower                 2      F       business     0   

     total_income                                   purpose  
12            NaN                         to have a wedding  
26            NaN                                 education  
29            NaN                    building a real estate  
41            NaN                  second-hand car purchase  
55            NaN                         to have a wedding  
65            NaN  transactions with commercial real estate  
67            NaN       purchase of the house for my family  
72            NaN  transactions with commercial real estate  
82            NaN                                   housing  
83            NaN                                   housing  
90            NaN                      housing transactions  
94            NaN                          having a wedding  
96            NaN               buy residential

In [ ]:

df_null = df_null.shape[0] / df.shape[0] 
print(f'Distribusi nilai yang hilang sebesar: {df_null:0%}')

Distribusi nilai yang hilang sebesar: 10.099884%


In [ ]:
df_null = df[(df['days_employed'].isnull()) & (df['total_income'].isnull())]

Possible causes of missing values in the data:
Looking at the distribution of the data, from a total of 2174 rows or about 10% of the total 21252 rows, where the missing values are not only in certain income_type or purpose or family_status, but spread out. I see that the missing values seem to be random.




In [ ]:


df.isna().sum() / df.shape[0] * 100

children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

There is a similarity between the original dataset and the filtered table distribution. This means that the missing values are random, not dependent on just 1 or 2 columns. 



In [ ]:

df_null['income_type'].value_counts(normalize=True)*100

employee         50.827967
business         23.367065
retiree          18.997240
civil servant     6.761730
entrepreneur      0.045998
Name: income_type, dtype: float64

In [ ]:
df_null['purpose'].value_counts(normalize=True)*100

having a wedding                            4.231831
to have a wedding                           3.725851
wedding ceremony                            3.495860
construction of own property                3.449862
housing transactions                        3.403864
buy real estate                             3.311868
transactions with my real estate            3.265869
purchase of the house for my family         3.265869
housing renovation                          3.219871
transactions with commercial real estate    3.219871
buy commercial real estate                  3.081877
buying property for renting out             2.989880
property                                    2.851886
real estate transactions                    2.805888
buy residential real estate                 2.805888
housing                                     2.759890
building a property                         2.713891
cars                                        2.621895
going to university                         2.

In [ ]:
df_null['family_status'].value_counts(normalize=True)*100

married              56.899724
civil partnership    20.331187
unmarried            13.247470
divorced              5.151794
widow / widower       4.369825
Name: family_status, dtype: float64

In [ ]:
df_null['children'].value_counts(normalize=True)*100

 0     66.191352
 1     21.849126
 2      9.383625
 3      1.655934
 20     0.413983
 4      0.321987
-1      0.137994
 5      0.045998
Name: children, dtype: float64

Looking at some of the columns where there are missing values in the days_employed and total_income columns, the ones with the most missing values are the productive group BUT no children or only 1 child. 
most missing values are productive groups BUT have no children or only 1 child.
The purpose of people applying for credit is dominated by wedding and property issues. 
By looking at the things above, it seems to be patterned, but the basis of the missing values in the days_employed and total_income columns is not certain. 
My temporary guess: days_employed and total_income are sensitive data that most people are reluctant to disclose even to the lender. The existence of data manipulation to achieve goals is also often the reason for lenders to choose to analyze the current account of a prospective debtor's checking or savings account. 

In [ ]:
#df.loc[df['days_employed'].isna(),'education'].str.lower().unique()

In [ ]:
#df['education'].str.lower().unique()

In [ ]:
#df['education'].str.lower().value_counts(normalize=True)

In [ ]:
#df['purpose'].value_counts(normalize=True).reset_index()

In [ ]:
#df.loc[df['total_income'].isna(),'purpose'].unique().shape

In [ ]:

df_null['gender'].value_counts(normalize=True)*100

F    68.26127
M    31.73873
Name: gender, dtype: float64


The missing values in days_employed and total_income columns can be described as follows:
Credit applicants are predominantly female (almost 70%) with married or civil partnership status where they have jobs mainly as employees, with the purpose mainly for wedding (probably for civil partnership) and property (probably for both married and civil partnership). 



In [ ]:

df['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [ ]:

df['education']=df['education'].str.lower()

In [ ]:

df['children'].value_counts(normalize=True)*100

 0     65.732869
 1     22.383275
 2      9.547038
 3      1.533101
 20     0.353078
-1      0.218351
 4      0.190476
 5      0.041812
Name: children, dtype: float64

In [ ]:
df['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [ ]:
df[(df['children']==-1)|(df['children']==20)][['children','income_type','purpose']].value_counts()
    

children  income_type  purpose                                 
 20       employee     buying a second-hand car                    4
                       buy commercial real estate                  3
                       transactions with commercial real estate    3
                       to own a car                                3
          business     university education                        3
                                                                  ..
-1        employee     to have a wedding                           1
                       supplementary education                     1
                       second-hand car purchase                    1
                       real estate transactions                    1
 20       retiree      transactions with commercial real estate    1
Length: 94, dtype: int64

In [ ]:
#df.loc[(df['children']==-1)|(df['children']==20),'children']=df['children'].median()

In [ ]:
#df['children'].value_counts(normalize=True)

In [ ]:

df.loc[(df['children']==-1)|(df['children']==20),'children']=df['children'].median()

There is no data to support removing the number of children -1 and 20 EVEN though they are obvious to me as outliers.
So changing the -1 and 20 values to the median is a moderate decision that does not change the distribution of the data significantly. 

In [ ]:
df['children'].value_counts(normalize=True)*100

0.0    66.304297
1.0    22.383275
2.0     9.547038
3.0     1.533101
4.0     0.190476
5.0     0.041812
Name: children, dtype: float64

In [ ]:
#df = df[(df.children != -1) & (df.children != 20)]
#df['children'].value_counts()

In [ ]:

#df['children'].value_counts(normalize=True)

In [ ]:

df['days_employed']

0         -8437.673028
1         -4024.803754
2         -5623.422610
3         -4124.747207
4        340266.072047
             ...      
21520     -4529.316663
21521    343937.404131
21522     -2113.346888
21523     -3112.481705
21524     -1984.507589
Name: days_employed, Length: 21525, dtype: float64

In [ ]:
df['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [ ]:
df[df['days_employed']<0].shape

(15906, 12)

In [ ]:
df[(df['days_employed']<0)]['income_type'].value_counts(normalize=True)*100

employee                       62.957375
business                       28.775305
civil servant                   8.248460
entrepreneur                    0.006287
paternity / maternity leave     0.006287
student                         0.006287
Name: income_type, dtype: float64

In [ ]:
df[(df['days_employed']>0)]['income_type'].value_counts(normalize=True)*100

retiree       99.941945
unemployed     0.058055
Name: income_type, dtype: float64

In [ ]:
wrong_data_days = df[df['days_employed']<0]
wrong_data_days['days_employed'].reset_index(drop=True)

0       -8437.673028
1       -4024.803754
2       -5623.422610
3       -4124.747207
4        -926.185831
            ...     
15901   -2351.431934
15902   -4529.316663
15903   -2113.346888
15904   -3112.481705
15905   -1984.507589
Name: days_employed, Length: 15906, dtype: float64

In [ ]:
percentage_days = wrong_data_days['days_employed'].count() / df['days_employed'].count()
print("Persentase data yang bermasalah pada kolom 'days_employed' yaitu:{:.0%} dari keseluruhan data.".format(percentage_days))

Persentase data yang bermasalah pada kolom 'days_employed' yaitu:82% dari keseluruhan data.



It is likely that the value in days_employed is a code from credit scoring where minus values are given to productive groups such as employee and business, while positive values are given to retiree and unemployed. But for ease of analysis in this context, let's change all values in the days_employed column to positive. 

In [ ]:

df['days_employed'] = df['days_employed'].abs()

In [ ]:

df['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

In [ ]:
#min(df['days_employed'])

In [ ]:

df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [ ]:
df['dob_years'].median()

42.0

In [ ]:
df['dob_years'].value_counts(normalize=True)*100

35    2.866434
40    2.829268
41    2.819977
34    2.801394
38    2.778165
42    2.773519
33    2.699187
39    2.662021
31    2.601626
36    2.578397
44    2.541231
29    2.531940
30    2.508711
48    2.499419
37    2.494774
50    2.387921
43    2.383275
32    2.369338
49    2.360046
28    2.336818
45    2.308943
27    2.290360
56    2.262485
52    2.248548
47    2.229965
54    2.225319
46    2.206736
58    2.141696
57    2.137050
53    2.132404
51    2.081301
59    2.062718
55    2.058072
26    1.895470
60    1.751452
25    1.658537
61    1.649245
62    1.635308
63    1.249710
64    1.231127
24    1.226481
23    1.180023
65    0.901278
66    0.850174
22    0.850174
67    0.775842
21    0.515679
0     0.469222
68    0.459930
69    0.394890
70    0.301974
71    0.269454
20    0.236934
72    0.153310
19    0.065041
73    0.037166
74    0.027875
75    0.004646
Name: dob_years, dtype: float64

In [ ]:
df.loc[df['dob_years']==0].shape

(101, 12)

In [ ]:
df.loc[df['dob_years']==0]['income_type'].value_counts()

employee         55
retiree          20
business         20
civil servant     6
Name: income_type, dtype: int64

In [ ]:
df.loc[df['dob_years']==0]['family_status'].value_counts()

married              49
civil partnership    21
unmarried            16
divorced             10
widow / widower       5
Name: family_status, dtype: int64

In [ ]:
pd.DataFrame(df['dob_years'].value_counts(normalize=True).sort_index()).T

0        19        20        21        22      23        24  \
dob_years  0.004692  0.00065  0.002369  0.005157  0.008502  0.0118  0.012265   

                 25        26        27  ...        66        67        68  \
dob_years  0.016585  0.018955  0.022904  ...  0.008502  0.007758  0.004599   

                 69       70        71        72        73        74        75  
dob_years  0.003949  0.00302  0.002695  0.001533  0.000372  0.000279  0.000046  

[1 rows x 58 columns]

In [ ]:

df_days_employed_prob_percent = (df['dob_years'] == 0).sum() / df.shape[0]
print(f'Persentase data yang bermasalah adalah: {df_days_employed_prob_percent:.2%}')

Persentase data yang bermasalah adalah: 0.47%


In [ ]:
df.loc[(df['dob_years']==0),'dob_years']=df['dob_years'].median()

With the distribution of the average dob_years data evenly distributed in 0-2% (see value_counts) and the problematic data is 'only' <0.5%, I think placing the median value on the problematic data will not damage the overall data information and can be accounted for. 

In [ ]:

df.loc[(df['dob_years'].isna())].shape

(0, 12)

In [ ]:

df['family_status'].value_counts(normalize=True)*100


married              57.514518
civil partnership    19.405343
unmarried            13.068525
divorced              5.551684
widow / widower       4.459930
Name: family_status, dtype: float64

In [ ]:
df['family_status'].describe()

count       21525
unique          5
top       married
freq        12380
Name: family_status, dtype: object

In [ ]:
df.loc[(df['family_status'].isna())].shape

(0, 12)

In [ ]:

df['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [ ]:
df['gender'].value_counts(normalize=True)*100

F      66.137050
M      33.858304
XNA     0.004646
Name: gender, dtype: float64

In [ ]:
# XNA gender is just dropped because the distribution is less than 0.5% and there are many NaNs
df_gender_value = df.pivot_table(index='gender', values='family_status', columns='purpose', aggfunc= 'count')
df_gender_value

purpose  building a property  building a real estate  \
gender                                                 
F                      412.0                   417.0   
M                      208.0                   209.0   
XNA                      NaN                     NaN   

purpose  buy commercial real estate  buy real estate  \
gender                                                 
F                             447.0            395.0   
M                             217.0            228.0   
XNA                             NaN              1.0   

purpose  buy residential real estate  buying a second-hand car  \
gender                                                           
F                              407.0                     324.0   
M                              200.0                     155.0   
XNA                              NaN                       NaN   

purpose  buying my own car  buying property for renting out    car  \
gender                                                               
F                    338.0                            443.0  339.0   
M                    167.0                            210.0  156.0   
XNA                    NaN                              NaN    NaN   

purpose  car purchase  ...  supplementary education  to become educated  \
gender                 ...                                                
F               320.0  ...                    311.0               266.0   
M               142.0  ...                    151.0               146.0   
XNA               NaN  ...                      NaN                 NaN   

purpose  to buy a car  to get a supplementary education  to have a wedding  \
gender                                                                       
F               295.0                             304.0              506.0   
M               177.0                             143.0              268.0   
XNA               NaN                               NaN                NaN   

purpose  to own a car  transactions with commercial real estate  \
gender                                                            
F               313.0                                     427.0   
M               167.0                                     224.0   
XNA               NaN                                       NaN   

purpose  transactions with my real estate  university education  \
gender                                                            
F                                   408.0                 298.0   
M                                   222.0                 155.0   
XNA                                   NaN                   NaN   

purpose  wedding ceremony  
gender                     
F                   547.0  
M                   250.0  
XNA                   NaN  

[3 rows x 38 columns]

In [ ]:
df = df[(df.gender != 'XNA')]
df['gender'].value_counts()

F    14236
M     7288
Name: gender, dtype: int64

In [ ]:

df['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

In [ ]:
df['income_type'].value_counts(normalize=True)*100

employee                       51.658614
business                       23.620145
retiree                        17.914886
civil servant                   6.778480
unemployed                      0.009292
entrepreneur                    0.009292
paternity / maternity leave     0.004646
student                         0.004646
Name: income_type, dtype: float64

In [ ]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [ ]:

print(df.duplicated().sum())

df_duplicated_percent = df.duplicated().sum() / df.shape[0] 
print(f'Distribusi nilai duplikat sebesar: {df_duplicated_percent:.2%}')

72
Distribusi nilai duplikat sebesar: 0.33%


In [ ]:

df = df.drop_duplicates().reset_index(drop=True)


In [ ]:

df.duplicated().sum()

0

In [ ]:

df.head(20)

children  days_employed  dob_years            education  education_id  \
0        1.0    8437.673028       42.0    bachelor's degree             0   
1        1.0    4024.803754       36.0  secondary education             1   
2        0.0    5623.422610       33.0  secondary education             1   
3        3.0    4124.747207       32.0  secondary education             1   
4        0.0  340266.072047       53.0  secondary education             1   
5        0.0     926.185831       27.0    bachelor's degree             0   
6        0.0    2879.202052       43.0    bachelor's degree             0   
7        0.0     152.779569       50.0  secondary education             1   
8        2.0    6929.865299       35.0    bachelor's degree             0   
9        0.0    2188.756445       41.0  secondary education             1   
10       2.0    4171.483647       36.0    bachelor's degree             0   
11       0.0     792.701887       40.0  secondary education             1   
12       0.0            NaN       65.0  secondary education             1   
13       0.0    1846.641941       54.0         some college             2   
14       0.0    1844.956182       56.0    bachelor's degree             0   
15       1.0     972.364419       26.0  secondary education             1   
16       0.0    1719.934226       35.0  secondary education             1   
17       0.0    2369.999720       33.0    bachelor's degree             0   
18       0.0  400281.136913       53.0  secondary education             1   
19       0.0   10038.818549       48.0  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate  
15     18691.345         construction of own property  
16   



1. the letters in the education column are uniform
2. the -1 and 20 values in the children column have been changed to median values
3. the XNA value in the gender column has been dropped because there is no significance
4. duplicates have been removed to avoid repetition of calculations or double counting
5. the value of 0 in the dob_years column has been replaced by the median value 
6. the minus value in the days_employed column has become positive to avoid miscalculation

In [ ]:

#df.head()

In [ ]:
education_dictionary=df[['education', 'education_id']].drop_duplicates().reset_index(drop=True)
education_dictionary

education  education_id
0    bachelor's degree             0
1  secondary education             1
2         some college             2
3    primary education             3
4      graduate degree             4

Saya hapus df.head() 

In [ ]:
family_status_dictionary=df[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)
family_status_dictionary

family_status  family_status_id
0            married                 0
1  civil partnership                 1
2    widow / widower                 2
3           divorced                 3
4          unmarried                 4

In [ ]:
df['dob_years'].describe()

count    21452.000000
mean        43.469933
std         12.213723
min         19.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [ ]:
df['dob_years'].value_counts()

42.0    696
35.0    616
40.0    607
41.0    605
34.0    601
38.0    597
33.0    581
39.0    572
31.0    559
36.0    554
44.0    545
29.0    544
30.0    537
48.0    536
37.0    536
50.0    513
43.0    512
32.0    509
49.0    508
28.0    503
45.0    496
27.0    493
52.0    484
56.0    483
47.0    477
54.0    476
46.0    472
53.0    459
57.0    456
58.0    454
51.0    446
55.0    443
59.0    443
26.0    408
60.0    374
25.0    357
61.0    354
62.0    348
63.0    269
24.0    263
64.0    260
23.0    252
65.0    193
22.0    183
66.0    182
67.0    167
21.0    111
68.0     99
69.0     85
70.0     65
71.0     56
20.0     51
72.0     33
19.0     14
73.0      8
74.0      6
75.0      1
Name: dob_years, dtype: int64

In [ ]:


def age_group(age):
    try:
        if age<30:
            return '<30'
        if 31<=age<=40:
            return '31-40'
        if 41<=age<=50:
            return '41-50'
        if 51<=age<=60:
            return '51-60'
        if 61<=age<=70:
            return '61-70'
        else:
            return '70+'
    except:
        return 0

In [ ]:

age_group(47)

'41-50'

In [ ]:

df['age_group']=df['dob_years'].apply(age_group)
df

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0  340266.072047       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21447       1.0    4529.316663       43.0  secondary education             1   
21448       0.0  343937.404131       67.0  secondary education             1   
21449       1.0    2113.346888       38.0  secondary education             1   
21450       3.0    3112.481705       38.0  secondary education             1   
21451       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21447  civil partnership                 1      F    business     0   
21448            married                 0      F     retiree     0   
21449  civil partnership                 1      M    employee     1   
21450            married                 0      M    employee     1   
21451            married                 0      F    employee     0   

       total_income                  purpose age_group  
0         40620.102    purchase of the house     41-50  
1         17932.802             car purchase     31-40  
2         23341.752    purchase of the house     31-40  
3         42820.568  supplementary education     31-40  
4         25378.572        to have a wedding     51-60  
...             ...                      ...       ...  
21447     35966.698     housing transactions     41-50  
21448     24959.969        purchase of a car     61-70  
21449     14347.610                 property     31-40  
21450     39054.888        buying my own car     31-40  
21451     13127.587             to buy a car     31-40  

[21452 rows x 13 columns]

In [ ]:

df_clean = df[df.notna()]
df_clean.head(20)

children  days_employed  dob_years            education  education_id  \
0        1.0    8437.673028       42.0    bachelor's degree             0   
1        1.0    4024.803754       36.0  secondary education             1   
2        0.0    5623.422610       33.0  secondary education             1   
3        3.0    4124.747207       32.0  secondary education             1   
4        0.0  340266.072047       53.0  secondary education             1   
5        0.0     926.185831       27.0    bachelor's degree             0   
6        0.0    2879.202052       43.0    bachelor's degree             0   
7        0.0     152.779569       50.0  secondary education             1   
8        2.0    6929.865299       35.0    bachelor's degree             0   
9        0.0    2188.756445       41.0  secondary education             1   
10       2.0    4171.483647       36.0    bachelor's degree             0   
11       0.0     792.701887       40.0  secondary education             1   
12       0.0            NaN       65.0  secondary education             1   
13       0.0    1846.641941       54.0         some college             2   
14       0.0    1844.956182       56.0    bachelor's degree             0   
15       1.0     972.364419       26.0  secondary education             1   
16       0.0    1719.934226       35.0  secondary education             1   
17       0.0    2369.999720       33.0    bachelor's degree             0   
18       0.0  400281.136913       53.0  secondary education             1   
19       0.0   10038.818549       48.0  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose age_group  
0      40620.102                purchase of the house     41-50  
1      17932.802                         car purchase     31-40  
2      23341.752                purchase of the house     31-40  
3      42820.568              supplementary education     31-40  
4      25378.572                    to have a wedding     51-60  
5      40922.170                purchase of the house       <30  
6      38484.156                 housing transactions     41-50  
7      21731.829                            education     41-50  
8      15337.093                     having a wedding     31-40  
9      23108.150  purchase of the house for my family     41-50  
10     18230.959                      buy real estate     31-40  
11     12331.077           buy commercial real estate     31-40  
12           NaN                    to have a wedding     61-70  
13     20873.317                 

In [ ]:

total_income_mean=df.pivot_table(index='age_group', values='total_income', columns=['family_status','income_type'], aggfunc='mean')
total_income_mean

family_status civil partnership                                           \
income_type            business civil servant      employee entrepreneur   
age_group                                                                  
31-40              32226.488634  27971.002465  26188.970341          NaN   
41-50              34674.779759  25880.232662  26348.367906          NaN   
51-60              29943.399358  27588.084179  25260.372509          NaN   
61-70              25344.320333  18470.537143  25969.447000          NaN   
70+                35580.018565  20298.131700  24875.619915          NaN   
<30                28728.988670  25854.964684  23844.296331    79866.103   

family_status                               divorced                \
income_type         retiree unemployed      business civil servant   
age_group                                                            
31-40          17911.417273        NaN  35826.226443  28837.787048   
41-50          30144.495914  32435.602  29484.902216  30684.786345   
51-60          21557.702697        NaN  36228.896422  22480.691000   
61-70          22465.054700        NaN  53331.198625  33647.509333   
70+            18497.191750        NaN  39187.578778  23823.804500   
<30            12807.071000        NaN  32032.629348  22972.571143   

family_status                              ...    married     unmarried  \
income_type        employee       retiree  ... unemployed      business   
age_group                                  ...                            
31-40          26197.694901  19028.182000  ...   9593.119  31837.347112   
41-50          27108.936197  25302.179222  ...        NaN  34701.375513   
51-60          24404.466161  21479.611913  ...        NaN  32310.710853   
61-70          27539.349944  22901.630785  ...        NaN  32760.650187   
70+            27081.081778  23331.503000  ...        NaN  35137.774077   
<30            23016.427966  12066.799000  ...        NaN  28757.204337   

family_status                                                      \
income_type   civil servant      employee       retiree   student   
age_group                                                           
31-40          30505.072171  25237.303057  29331.419200       NaN   
41-50          30724.945242  26717.145691  26557.224632       NaN   
51-60          25432.284000  27399.417723  23810.634551       NaN   
61-70          35849.229500  37546.384842  20325.937656       NaN   
70+            25056.889375  25765.274911  18989.549444       NaN   
<30            25592.389711  23826.669745  11227.035000  15712.26   

family_status widow / widower                                            
income_type          business civil servant      employee       retiree  
age_group                                                                
31-40            29943.028615  27335.301333  25827.418333  34632.356000  
41-50            27085.293556  26605.605083  23521.346857  21514.467300  
51-60            27150.825333  22853.840421  23587.947322  20913.732735  
61-70            33531.175500  30257.731667  25622.147108  21649.627017  
70+                       NaN  24525.224000  25674.349500  17795.521000  
<30                       NaN           NaN  22915.450000           NaN  

[6 rows x 25 columns]

In [ ]:

total_income_median=df.pivot_table(index='age_group', values='total_income', aggfunc='median')
total_income_median

total_income
age_group              
31-40        24825.1865
41-50        24563.6500
51-60        22056.7710
61-70        19705.8550
70+          23083.2645
<30          22735.9110

In [ ]:

df['total_income'] = df.groupby(['age_group'])['total_income'].transform(lambda x: x.fillna(x.median()))
        

In [ ]:

df.loc[(df['total_income'].isna())].shape

(0, 13)

In [ ]:

df.isna().sum()

children               0
days_employed       2102
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
age_group              0
dtype: int64

In [ ]:

df.describe()


children  days_employed     dob_years  education_id  \
count  21452.000000   19350.000000  21452.000000  21452.000000   
mean       0.471331   66918.065141     43.469933      0.817033   
std        0.751093  139033.698578     12.213723      0.548639   
min        0.000000      24.141633     19.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     42.000000      1.000000   
75%        1.000000    5538.423086     53.000000      1.000000   
max        5.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21452.000000  21452.000000   21452.000000  
mean           0.973895      0.081158   26446.496403  
std            1.421634      0.273084   15690.170869  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   17217.441750  
50%            0.000000      0.000000   23234.038000  
75%            1.000000      0.000000   31328.693750  
max            4.000000      1.000000  362496.645000

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21452 entries, 0 to 21451
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21452 non-null  float64
 1   days_employed     19350 non-null  float64
 2   dob_years         21452 non-null  float64
 3   education         21452 non-null  object 
 4   education_id      21452 non-null  int64  
 5   family_status     21452 non-null  object 
 6   family_status_id  21452 non-null  int64  
 7   gender            21452 non-null  object 
 8   income_type       21452 non-null  object 
 9   debt              21452 non-null  int64  
 10  total_income      21452 non-null  float64
 11  purpose           21452 non-null  object 
 12  age_group         21452 non-null  object 
dtypes: float64(4), int64(3), object(6)
memory usage: 2.1+ MB


In [ ]:

days_employed_median=df.pivot_table(index='age_group', values='days_employed', aggfunc='median')
days_employed_median

days_employed
age_group               
31-40        1630.193189
41-50        2199.733119
51-60        6481.039027
61-70      356081.013729
70+          1799.537390
<30           998.961907

In [ ]:

days_employed_mean=df.pivot_table(index='age_group', values='days_employed', aggfunc='mean')
days_employed_mean

days_employed
age_group               
31-40        4691.122800
41-50       17693.274606
51-60      153165.111212
61-70      288992.710281
70+         55910.432052
<30          2082.398127

The median makes more sense because the distribution of the data (weighted scores) is not clearly known (it would be easier with a visual graph).

In [ ]:

df['days_employed'] = df.groupby(['age_group'])['days_employed'].transform(lambda x: x.fillna(x.median()))


In [ ]:
df['days_employed'].describe()

count     21452.000000
mean      63985.511143
std      136316.866892
min          24.141633
25%         998.961907
50%        2199.733119
75%        5798.185967
max      401755.400475
Name: days_employed, dtype: float64

In [ ]:
# Periksa apakah fungsi Anda dapat bekerja
df['days_employed'].head(50)


0       8437.673028
1       4024.803754
2       5623.422610
3       4124.747207
4     340266.072047
5        926.185831
6       2879.202052
7        152.779569
8       6929.865299
9       2188.756445
10      4171.483647
11       792.701887
12    356081.013729
13      1846.641941
14      1844.956182
15       972.364419
16      1719.934226
17      2369.999720
18    400281.136913
19     10038.818549
20      1311.604166
21       253.685166
22      1766.644138
23       272.981385
24    338551.952911
25    363548.489348
26      2199.733119
27       529.191635
28       717.274324
29    356081.013729
30    335581.668515
31      1682.083438
32      4649.910832
33      1548.637544
34      4488.067031
35    394021.072184
36       176.216688
37      6448.810860
38       597.881827
39       650.587796
40      1030.362606
41      2199.733119
42      1257.496190
43      4375.681384
44      1362.041728
45      1039.451741
46      2262.712304
47      2689.137274
48      3341.067886
49      1181.443228


In [ ]:

df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_group           0
dtype: int64

In [ ]:

df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_group           0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21452 entries, 0 to 21451
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21452 non-null  float64
 1   days_employed     21452 non-null  float64
 2   dob_years         21452 non-null  float64
 3   education         21452 non-null  object 
 4   education_id      21452 non-null  int64  
 5   family_status     21452 non-null  object 
 6   family_status_id  21452 non-null  int64  
 7   gender            21452 non-null  object 
 8   income_type       21452 non-null  object 
 9   debt              21452 non-null  int64  
 10  total_income      21452 non-null  float64
 11  purpose           21452 non-null  object 
 12  age_group         21452 non-null  object 
dtypes: float64(4), int64(3), object(6)
memory usage: 2.1+ MB


In [ ]:

df['purpose'].value_counts()


wedding ceremony                            791
having a wedding                            767
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house                       646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
purchase of my own house                    620
buy real estate                             620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

In [ ]:

df['purpose'].sort_values().unique()

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

In [ ]:

def purpose_common(purpose):
    
    if 'property' in purpose:
        return 'property'
    elif 'estate' in purpose:
        return 'property'
    elif 'hous' in purpose:
        return 'property'
    elif 'car' in purpose:
        return 'car'
    elif 'educ' in purpose:
        return 'education'
    elif 'univ' in purpose:
        return 'education'
    elif 'wedd' in purpose:
        return 'wedding'

In [ ]:

df['general_purpose'] = df['purpose'].apply(purpose_common)
print(df['general_purpose'].value_counts())
print()
df['general_purpose'].count()


property     10810
car           4306
education     4013
wedding       2323
Name: general_purpose, dtype: int64



21452

In [ ]:

df

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0  340266.072047       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21447       1.0    4529.316663       43.0  secondary education             1   
21448       0.0  343937.404131       67.0  secondary education             1   
21449       1.0    2113.346888       38.0  secondary education             1   
21450       3.0    3112.481705       38.0  secondary education             1   
21451       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21447  civil partnership                 1      F    business     0   
21448            married                 0      F     retiree     0   
21449  civil partnership                 1      M    employee     1   
21450            married                 0      M    employee     1   
21451            married                 0      F    employee     0   

       total_income                  purpose age_group general_purpose  
0         40620.102    purchase of the house     41-50        property  
1         17932.802             car purchase     31-40             car  
2         23341.752    purchase of the house     31-40        property  
3         42820.568  supplementary education     31-40       education  
4         25378.572        to have a wedding     51-60         wedding  
...             ...                      ...       ...             ...  
21447     35966.698     housing transactions     41-50        property  
21448     24959.969        purchase of a car     61-70             car  
21449     14347.610                 property     31-40        property  
21450     39054.888        buying my own car     31-40             car  
21451     13127.587             to buy a car     31-40             car  

[21452 rows x 14 columns]

In [ ]:
def income_class(total_income):
 
    if total_income <= 32000:
        return 'poor'
    elif total_income <= 53000:
        return 'lower middle class'
    elif total_income <= 106000:
        return 'middle class'
    elif total_income <= 373000:
        return 'upper middle class'
    else:
        return 'rich'

I categorized total_income into several category classes based on the economic level classification, namely poor, lower middle class, middle class, upper middle class and rich. This is intended to categorize the categories only so as to facilitate the decision-making process.

In [ ]:
df['economic_class'] = df['total_income'].apply(income_class)

In [ ]:
df['economic_class'].value_counts()



poor                  16387
lower middle class     3985
middle class           1000
upper middle class       80
Name: economic_class, dtype: int64

In [ ]:

df

children  days_employed  dob_years            education  education_id  \
0           1.0    8437.673028       42.0    bachelor's degree             0   
1           1.0    4024.803754       36.0  secondary education             1   
2           0.0    5623.422610       33.0  secondary education             1   
3           3.0    4124.747207       32.0  secondary education             1   
4           0.0  340266.072047       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21447       1.0    4529.316663       43.0  secondary education             1   
21448       0.0  343937.404131       67.0  secondary education             1   
21449       1.0    2113.346888       38.0  secondary education             1   
21450       3.0    3112.481705       38.0  secondary education             1   
21451       2.0    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21447  civil partnership                 1      F    business     0   
21448            married                 0      F     retiree     0   
21449  civil partnership                 1      M    employee     1   
21450            married                 0      M    employee     1   
21451            married                 0      F    employee     0   

       total_income                  purpose age_group general_purpose  \
0         40620.102    purchase of the house     41-50        property   
1         17932.802             car purchase     31-40             car   
2         23341.752    purchase of the house     31-40        property   
3         42820.568  supplementary education     31-40       education   
4         25378.572        to have a wedding     51-60         wedding   
...             ...                      ...       ...             ...   
21447     35966.698     housing transactions     41-50        property   
21448     24959.969        purchase of a car     61-70             car   
21449     14347.610                 property     31-40        property   
21450     39054.888        buying my own car     31-40             car   
21451     13127.587             to buy a car     31-40             car   

           economic_class  
0      lower middle class  
1                    poor  
2                    poor  
3      lower middle class  
4                    poor  
...                   ...  
21447  lower middle class  
21448                poor  
21449                poor  
21450  lower middle class  
21451                poor  

[21452 rows x 15 columns]

In [ ]:

pivot_table_children = df.pivot_table(index='children', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_children

debt            0       1
children                 
0.0       13140.0  1072.0
1.0        4364.0   444.0
2.0        1858.0   194.0
3.0         303.0    27.0
4.0          37.0     4.0
5.0           9.0     NaN

In [ ]:
df['debt'].value_counts()

0    19711
1     1741
Name: debt, dtype: int64

In [ ]:
pivot_table_children['percentage'] = pivot_table_children[1] / (pivot_table_children[1] + pivot_table_children[0]) * 100
pivot_table_children

debt            0       1  percentage
children                             
0.0       13140.0  1072.0    7.542921
1.0        4364.0   444.0    9.234609
2.0        1858.0   194.0    9.454191
3.0         303.0    27.0    8.181818
4.0          37.0     4.0    9.756098
5.0           9.0     NaN         NaN


1.prospective debtors who do not have children, will have a 'only' 7.5% chance of defaulting / being burdened with debt
2.prospective debtors who have 1 to 4 children will have a chance of defaulting 8-9.7% 

In [ ]:

pivot_table_family_status = df.pivot_table(index='family_status', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_family_status



debt                   0    1
family_status                
civil partnership   3761  388
divorced            1110   85
married            11408  931
unmarried           2536  274
widow / widower      896   63

In [ ]:
pivot_table_family_status['percentage'] = pivot_table_family_status[1] / (pivot_table_family_status[1] + pivot_table_family_status[0]) * 100
pivot_table_family_status

debt                   0    1  percentage
family_status                            
civil partnership   3761  388    9.351651
divorced            1110   85    7.112971
married            11408  931    7.545182
unmarried           2536  274    9.750890
widow / widower      896   63    6.569343


1. prospective debtors with the condition of unmarried and civil partnership (spouse) will have the highest probability of default (9.8% and 9.4%)
2. while prospective debtors with widow/widower conditions 'only' have the lowest chance of default (7%)



In [ ]:

pivot_table_economic_class = df.pivot_table(index='economic_class', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_economic_class


debt                    0     1
economic_class                 
lower middle class   3704   281
middle class          928    72
poor                15004  1383
upper middle class     75     5

In [ ]:
pivot_table_economic_class['percentage'] = pivot_table_economic_class[1] / (pivot_table_economic_class[1] + pivot_table_economic_class[0]) * 100
pivot_table_economic_class


debt                    0     1  percentage
economic_class                             
lower middle class   3704   281    7.051443
middle class          928    72    7.200000
poor                15004  1383    8.439617
upper middle class     75     5    6.250000

1. the more the category is 'poor', the greater the chance of default.

In [ ]:

pivot_table_general_purpose = df.pivot_table(index='general_purpose', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_general_purpose['percentage'] = pivot_table_general_purpose[1] / (pivot_table_general_purpose[1] + pivot_table_general_purpose[0]) * 100
pivot_table_general_purpose


debt                 0    1  percentage
general_purpose                        
car               3903  403    9.359034
education         3643  370    9.220035
property         10028  782    7.234043
wedding           2137  186    8.006888


1. the highest probability of default based on the data above, it can be concluded that prospective debtors with the aim of buying or renting a car and or financing education (>9%)
2. while those who aim to buy or rent property have the lowest chance of default. 




Conclusion:
1. There are missing values in this dataset that exist randomly without a pattern. In overcoming this missing data, it is done by filling in the data from the median value of the available data.
2. Duplicate data in this dataset is fixed by changing the way of writing to lowercase letters.
3. There is also a strange data value in the 'children' column that has children totaling -1 and 20. The data is fixed by changing the negative value to positive and not deleting the existing data value due to the lack of convincing information to delete the data value. 
4. In the 'dob_years' column, there was an odd data value of 0. The data was corrected by replacing the value with the median of 42. 
5. The new corrected data is checked for duplicates. The duplicate drop process is performed.
6. The dictionary data is dropped because it is an easy way to fix the data.
7. The missing value in the 'total_income' column is fixed by creating an age category for the customer. Then filled with the median value of 'total_income'.
8. Missing values in the 'days_employed' column are fixed by filling in the blank values with the available median data.
9. Data categorization to facilitate decision making and hypothesis checking.
10. From the results of the hypothesis checking data, it is known that the probability of customers having children, family status, total income, and purpose of purchase (can be said not to affect customer default because it is known in the data that 90% and above of customers successfully pay the loan.
